bind: misc fixes and cleanups

Alphix opened the pull request:
## 📦 Package Details

**Maintainer:** @nmeyerhans
<sub>(You can find this by checking the history of the package `Makefile`.)</sub>

**Description:**
<!-- Briefly describe what this package does or what changes are introduced -->
This patch series refreshes the bind packaging to take recent developments and defaults in the upstream bind project into account. A lot of things that used to be necessary (like empty zones for localhost forward and reverse zones) are now handled by default. Also, outdated root hints and trust anchors have been updated or removed. The result is a much simpler package. Built, installed and tested on my router (running 24.10).

After package installation with this patch series applied:
```
root@router:~# ls -al /etc/bind/
drwxr-xr-x    2 root     root          1024 Jul 27 04:29 .
drwxr-xr-x    1 root     root          1024 Jul 27 04:27 ..
-rw-r--r--    1 root     root          3310 Jul 27 04:15 db.root
-rw-r--r--    1 root     root          1330 Jul 27 04:29 named.conf
-rw-------    1 bind     bind           100 Jul 27 04:27 rndc.key
```

---

## 🧪 Run Testing Details

- **OpenWrt Version:** 24.10
- **OpenWrt Target/Subtarget:** mvebu/cortexa9
- **OpenWrt Device:** cznic_turris-omnia

---

## ✅ Formalities

- [x] I have reviewed the [CONTRIBUTING.md](../CONTRIBUTING.md) file for detailed contributing guidelines.

### If your PR contains a patch:

- [x] It can be applied using `git am`
- [x] It has been refreshed to avoid offsets, fuzzes, etc., using
  ```bash
  make package/<your-package>/refresh V=s
  ```
- [x] It is structured in a way that it is potentially upstreamable **Not applicable**
<sub>(e.g., subject line, commit description, etc.)</sub>
<sub>We must try to upstream patches to reduce maintenance burden.</sub>

Alphix commented:
Ok, bind9 builds for me on 24.10, but it doesn't build on `master` with or without my patches.

The build already dies on the configure stage (without my patches applied), because the autoconf magic manages to mix up the host gcc and the cross-compiling gcc:

```
checking build system type... x86_64-pc-linux-gnu
checking host system type... mipsel-openwrt-linux-gnu
checking target system type... mipsel-openwrt-linux-gnu
...
checking whether the compiler supports GNU C... yes
checking whether mipsel-openwrt-linux-musl-gcc accepts -g... yes
checking for mipsel-openwrt-linux-musl-gcc option to enable C23 features... -std=gnu23
checking whether mipsel-openwrt-linux-musl-gcc -std=gnu23 understands -c and -o together... yes
...
checking for mipsel-openwrt-linux-gcc... (cached) mipsel-openwrt-linux-musl-gcc
checking whether the compiler supports GNU C... (cached) yes
checking whether mipsel-openwrt-linux-musl-gcc accepts -g... (cached) yes
checking for mipsel-openwrt-linux-musl-gcc option to enable C23 features... (cached) -std=gnu23
checking whether mipsel-openwrt-linux-musl-gcc -std=gnu23 understands -c and -o together... (cached) yes
checking dependency style of mipsel-openwrt-linux-musl-gcc -std=gnu23... (cached) none
checking how to run the C preprocessor... mipsel-openwrt-linux-musl-gcc -std=gnu23 -E
checking for x86_64-pc-linux-gnu-gcc... no
checking for gcc... gcc
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C23 features... (cached) -std=gnu23
checking whether gcc -std=gnu23 understands -c and -o together... no
checking dependency style of /builder/build_dir/target-mipsel_24kc_musl/bind-9.20.11/compile gcc -std=gnu23... none
checking whether the C compiler works... no
configure: error: in '/builder/build_dir/target-mipsel_24kc_musl/bind-9.20.11':
configure: error: C compiler cannot create executables
See 'config.log' for more details
```

Ok, so I figured out that removing `PKG_FIXUP:=autoreconf` from the `Makefile` allowed the configure step to complete, but compilation still chokes on `libtool` magic:

```
make[7]: Entering directory '/home/build/openwrt/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/bind-9.20.11/bin/rndc'
...
/bin/bash ../../libtool  --tag=CC   --mode=link arm-openwrt-linux-muslgnueabi-gcc -Wall -Wextra -Wwrite-strings -Wpointer-arith -Wno-missing-field-initializers -Wformat -Wshadow -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=format-security -Werror=parentheses -Werror=implicit -Werror=strict-prototypes -Werror=vla -fno-strict-aliasing -fno-delete-null-pointer-checks -fdiagnostics-show-option -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -mfloat-abi=hard -fmacro-prefix-map=/home/build/openwrt/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/bind-9.20.11=bind-9.20.11 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro   -DHAVE_JSON_C -UHAVE_JSON  -I/home/build/openwrt/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/include -pthread -Wl,--export-dynamic  -L/home/build/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.3.0_musl_eabi/usr/lib -L/home/build/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.3.0_musl_eabi/lib -fuse-ld=bfd -znow -zrelro   -Wl,--gc-sections,--as-needed  -o rndc rndc.o util.o ../../lib/isc/libisc.la ../../lib/dns/libdns.la ../../lib/isccc/libisccc.la ../../lib/isccfg/libisccfg.la -lpthread  -L/home/build/openwrt/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/lib -lurcu -lurcu-common -lurcu-cds
libtool: link: arm-openwrt-linux-muslgnueabi-gcc -Wall -Wextra -Wwrite-strings -Wpointer-arith -Wno-missing-field-initializers -Wformat -Wshadow -Werror=implicit-function-declaration -Werror=missing-prototypes -Werror=format-security -Werror=parentheses -Werror=implicit -Werror=strict-prototypes -Werror=vla -fno-strict-aliasing -fno-delete-null-pointer-checks -fdiagnostics-show-option -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -mfloat-abi=hard -fmacro-prefix-map=/home/build/openwrt/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/bind-9.20.11=bind-9.20.11 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z -Wl,now -Wl,-z -Wl,relro -DHAVE_JSON_C -UHAVE_JSON -I/home/build/openwrt/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/include -Wl,--export-dynamic -fuse-ld=bfd -znow -zrelro -Wl,--gc-sections -Wl,--as-needed -o .libs/rndc rndc.o util.o  -L/home/build/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.3.0_musl_eabi/usr/lib -L/home/build/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.3.0_musl_eabi/lib ../../lib/isc/.libs/libisc.so ../../lib/dns/.libs/libdns.so ../../lib/isccc/.libs/libisccc.so ../../lib/isccfg/.libs/libisccfg.so -lpthread -L/home/build/openwrt/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/lib -lurcu -lurcu-common -lurcu-cds -pthread
/home/build/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.3.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/14.3.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld.bfd: warning: libns-9.20.11.so, needed by ../../lib/isccfg/.libs/libisccfg.so, not found (try using -rpath or -rpath-link)
/home/build/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.3.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/14.3.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld.bfd: ../../lib/isccfg/.libs/libisccfg.so: undefined reference to `ns_plugin_check'
/home/build/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-14.3.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/14.3.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld.bfd: ../../lib/isccfg/.libs/libisccfg.so: undefined reference to `ns_plugin_expandpath'
collect2: error: ld returned 1 exit status
```

nmeyerhans commented on net/bind/Makefile:
```diff
@@ -24,9 +24,6 @@ PKG_SOURCE_URL:= \
 	https://ftp.isc.org/isc/bind9/$(PKG_VERSION)
 PKG_HASH:=4da2d532e668bc21e883f6e6d9d3d81794d9ec60b181530385649a56f46ee17a
 
-PKG_FIXUP:=autoreconf
-PKG_REMOVE_FILES:=aclocal.m4 libtool.m4
```
It would be nice to understand why this is suddenly breaking the build, but I can confirm that it builds ok with these lines removed and fails when they're present.

Alphix replied:
I had the impression that the `autoreconf` fixup fails to actually fixup the autotool mess in bind9 (I seem to recall that there's like 8k lines of Makefiles and magic....but it's at least been replaced with meson in the development branch of bind). IIRC there were one or more '.rej' files in the build dir (which would indicate a failed attempt at patching files)

nmeyerhans commented on net/bind/files/named.init:
```diff
@@ -3,86 +3,48 @@
-
-    rndc-confgen > $rndc_temp
-
-    sed -r -n \
-	-e '/^# options \{$/,/^\};$/{ s/^/# / }' \
-	-e p \
-	-e '/^# End of rndc\.conf$/q' \
-	< $rndc_temp > $rndc_conf_file
-
-    sed -r -n \
-	-e '1,/^# End of rndc\.conf$/ { b done }' \
-	-e '/^# Use with the following in named.conf/ { p ; b done }' \
-	-e '/^# End of named\.conf$/ { p ; b done }' \
-	-e '/^# key /,$ { s/^# // ; p }' \
-	-e ': done' \
-	< $rndc_temp > $named_options_file
```
Happy to see this go. :+1:

Alphix replied:
I'm happy that you're happy :D

nmeyerhans commented on net/bind/files/bind/named.conf:
```diff
@@ -4,8 +4,6 @@ options {
 	directory "/var/cache/bind";
 };
 
-include "/tmp/bind/named.conf.local";
```
This comes from 52c4c84700c7b5ecc80099a197d44cfbf4e9cf17. It's probably worth keeping around since it provides an easy mechanism to add additional zone configuration at runtime, while keeping it in `/tmp/` makes it clear that we're not talking about persisting state long-term.

Alphix replied:
Hm...I see. How about putting it in `/var/cache/bind` though? That's kind of the standard location for `bind` files that are temporary in nature (like the dynamic zones that https://github.com/openwrt/packages/commit/52c4c84700c7b5ecc80099a197d44cfbf4e9cf17 refers to)...

On the other hand...how "dynamic" are the zones are we talking about? If it's about the *content* of the zone, then that anyway doesn't end up in `/tmp/bind/named.conf.local` or whatever...it would get written to some other file and referred to from an appropriate zone block (`zone "blah.lan" { type master; file "/tmp/blah.db"; }};`) in the config file....? Or are we really talking about a large bunch of zones that would appear/disappear wholesale from one reboot to the next? Seems a bit....far-fetched?

nmeyerhans replied:
I can't really speak to how any of this is actually being used, as I'm not making use of this feature myself. In terms of the implementation, on OpenWRT either path will be on the same filesystem, so it would be safe. However, it's technically a public interface, so changing the path can potentially break people's configuration.

I'm not aware of any packages within OpenWRT itself that make use of this path, but it's been there for 5 years, so it has had time to get adopted.  If we do change the path, I'd want to keep the commit in `master` and not backport it to any of the existing stable branches (24.10, etc).

Alphix replied:
> I'm not aware of any packages within OpenWRT itself that make use of this path, but it's been there for 5 years, so it has had time to get adopted. If we do change the path, I'd want to keep the commit in `master` and not backport it to any of the existing stable branches (24.10, etc).

Perfectly fine by me... 😄

pprindeville replied:
> I can't really speak to how any of this is actually being used, as I'm not making use of this feature myself. In terms of the implementation, on OpenWRT either path will be on the same filesystem, so it would be safe. However, it's technically a public interface, so changing the path can potentially break people's configuration.
> 
> I'm not aware of any packages within OpenWRT itself that make use of this path, but it's been there for 5 years, so it has had time to get adopted. If we do change the path, I'd want to keep the commit in `master` and not backport it to any of the existing stable branches (24.10, etc).

It’s used by ISC-DHCP (and Kea if I can ever get my libubox changes adopted). You should be able to search GitHub for that path.

nmeyerhans commented:
You have a typo in the commit message for 52c4c84700c7b5ecc80099a197d44cfbf4e9cf17

> First, change the "directory" to point to /etc/init.d/named

`/etc/init.d/named` should be `/var/cache/bind`

Looks good in general, and is so far OK in my testing on `master`.

Alphix commented:
> You have a typo in the commit message for [52c4c84](https://github.com/openwrt/packages/commit/52c4c84700c7b5ecc80099a197d44cfbf4e9cf17)

Well, that's not my commit :)

But yeah, I see what you mean, and I pushed a rebase where the fixed commit message is the only change.

vochong commented:
@Alphix 

I cherry-picked your PR and the build worked fine for several different models.  Thanks.

Alphix commented:
@nmeyerhans I'm thinking I should add one more patch, making `/var/lib/bind` a symlink to `/etc/bind/zones` (and document the two locations very briefly in `/etc/bind/named.conf`). That way, everything that should survive reboots/upgrades is below `/etc/bind` but administrators can still work with the standard paths. Don't know if you'd prefer a separate patch or if I should add it to this PR.

pprindeville commented on net/bind/files/bind/db.root:
```diff
@@ -1,90 +1,92 @@
 K.ROOT-SERVERS.NET.      3600000      A     193.0.14.129
 K.ROOT-SERVERS.NET.      3600000      AAAA  2001:7fd::1
-;
+; 
```
Please unfollow the white space added to the ends of lines.

Alphix replied:
That's how the file is when downloaded from the canonical source...I think it makes sense to keep the file unadulterated?

pprindeville commented on net/bind/Makefile:
```diff
@@ -221,30 +205,15 @@ define Package/bind-server/install
 endef
 
 define Package/bind-server/conffiles
-/etc/bind/db.0
```
Why are these going away?

Alphix replied:
They're built into recent versions of `bind`

pprindeville commented on net/bind/files/bind/db.root:
```diff
@@ -1,90 +1,92 @@
 M.ROOT-SERVERS.NET.      3600000      A     202.12.27.33
 M.ROOT-SERVERS.NET.      3600000      AAAA  2001:dc3::35
-; End of file
+; End of file
```
Ditto

pprindeville reviewed (changes requested):
Minor changes requested

pprindeville commented:
@nmeyerhans On the subject of ISC-DHCP, I'm trying to deprecate it (it's EOL'd) and I have a backend that takes the current UCI and generates a Kea JSON file, but it requires a few patches to `libubox` I've been trying to get reviewed/merged for 2 years now.  They're pretty simple, innocuous changes but in the usual Openwrt style, sometimes things unexpected fall into a bottomless pit of silence.

If you have any sway in getting this unblocked, I'd appreciate it.

https://lists.openwrt.org/pipermail/openwrt-devel/2025-May/043961.html
https://lists.openwrt.org/pipermail/openwrt-devel/2025-May/043962.html
https://lists.openwrt.org/pipermail/openwrt-devel/2025-May/043963.html
https://lists.openwrt.org/pipermail/openwrt-devel/2025-May/043964.html
https://lists.openwrt.org/pipermail/openwrt-devel/2025-May/043965.html
https://lists.openwrt.org/pipermail/openwrt-devel/2025-May/043960.html
https://lists.openwrt.org/pipermail/openwrt-devel/2025-May/043966.html

One of these is even a trivial one line fix that Felix himself proposed, so I struggle to understand why there's no easy way to get these merged.

Alphix commented:
> There is a tight coupling between ISC-DHCP and Bind. Please make sure that dynamic updates aren’t broken.

Sorry about that. I've pushed a commit which sets `allow-new-zones yes;` in `named.conf` and added a separate PR (#27153) to make `isc-dhcp` use this functionality.

nmeyerhans commented:
> @nmeyerhans I'm thinking I should add one more patch, making `/var/lib/bind` a symlink to `/etc/bind/zones` (and document the two locations very briefly in `/etc/bind/named.conf`). That way, everything that should survive reboots/upgrades is below `/etc/bind` but administrators can still work with the standard paths. Don't know if you'd prefer a separate patch or if I should add it to this PR.

Let's keep that separate so we can merge this PR without further delay.

Let's make sure to coordinate merging of this PR and #27153 so we don't break anything.

Alphix commented on net/bind/files/named.init:
```diff
@@ -3,86 +3,55 @@
+run_dir=/var/run/named
+log_dir=/var/log/named
+cache_dir=/var/cache/bind
+zone_dir=/etc/bind/zones
```
Ok

Alphix commented on net/bind/files/named.init:
```diff
@@ -3,86 +3,55 @@
+	done
+
+	if [ ! -e $lib_dir ]; then
+		mkdir -p $(dirname $lib_dir) > /dev/null 2>&1 || true
```
Ok, I've removed the checks

Alphix commented on net/bind/files/named.init:
```diff
@@ -3,86 +3,55 @@
+		chmod 0640 /etc/bind/rndc.key
+	fi
+
+	if [ -z "$(ip -6 -o route show default)" ]; then
```
I doubt it'll end up being a test in the end. Rather it'll be something like:

```
config_load 'bind'
config_get args global args
...
procd_append_param command '${args:-}'
```

So I think it makes sense to leave it as-is for now, and deal with the final look of that line in whatever fix ends up being merged for #26327

nmeyerhans commented:
> > ...merge this PR without further delay.
> 
> @nmeyerhans gentle ping? I think all issues have been addressed, and if you're happy with the state of this PR, I can add the one commit from #27153 so that both `bind` and `isc-dhcp` are modified in lockstep...

@Alphix Yes, thank you, this looks good to me.

Alphix commented:
> > @nmeyerhans gentle ping? I think all issues have been addressed, and if you're happy with the state of this PR, I can add the one commit from #27153 so that both `bind` and `isc-dhcp` are modified in lockstep...
> 
> @Alphix Yes, thank you, this looks good to me.

@nmeyerhans Ok, rebased, added the patch from #27153 and reorganised things a little bit, but the end result is the same as before (+ a PKG_RELEASE bump). Please have a look.

pprindeville reviewed (approved):
LGTM

pprindeville commented:
> @pprindeville any idea what's up with the isc-dhcp build failure on riscv64_generic? Is this a new issue?

`riscv64_generic` is broken more often than not.  If the others are building and passing, that's good enough for me.

pprindeville commented:
No pressure, but it would be great to get this merged this weekend so I can pivot to the ISC-DHCP to Kea migration stuff.

pprindeville commented:
Awesome possum!
